### PR TITLE
Implement a basic functionality for add/remove node feature + a basic demo in Apollo

### PR DESCRIPTION
**Pull request by yontyon**
To support adding and removing a node from the cluster we take the following actions:
1. stop the replica on the very same checkpoint (wedge)
2. kill all replicas
3. load the replicas with the new configuration that contains/not-contains the nodes' details
4. restart the new replicas.

Effectively, this means we start a new blockchain with the state of the former blockchain.
Yet, as concord-bft saves all the configuration data in persistent storage, the first step would be to implement functionality to clear the metadata storage before shutting down a replica (see https://github.com/vmware/concord-bft/pull/876).

In this PR we introduce a working example for removing that metadata storage.

Edit: We also added a n-f wedge support. Notice that currently, this is not safe to use it in other reconfiguration action except for removing nodes.

### CODE REVIEW

File: tests/apollo/CMakeLists.txt [hunk 1]
```diff
@@ -72,7 +72,7 @@ foreach(STORAGE_TYPE ${STORAGE_TYPES})
           WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
 
   add_test(NAME skvbc_reconfiguration_${STORAGE_TYPE} COMMAND sh -c
-          "env ${APOLLO_TEST_ENV} STORAGE_TYPE=${STORAGE_TYPE} python3 -m unittest test_skvbc_reconfiguration ${TEST_OUTPUT}"
+          "env ${APOLLO_TEST_ENV} LEAKCHECK=ON STORAGE_TYPE=${STORAGE_TYPE} python3 -m unittest test_skvbc_reconfiguration ${TEST_OUTPUT}"
```
Comment (yontyon) [inline tests/apollo/CMakeLists.txt:75]: I've added the LEAKCHECK=ON because, in order to clean the metadata storage, the replicas need to be shutdown normally.
Apollo is calling to p.kill unless LEACKCHECK=ON and then it calls to p.terminate

Comment (teoparvanov) [reply tests/apollo/CMakeLists.txt:75]: @HristoStaykov what do you think about this?

Comment (HristoStaykov) [reply tests/apollo/CMakeLists.txt:75]: I added this because for LeakSanitizer reports we needed to terminate the Replica processes with SIGTERM instead of SIGKILL. This adds slow down for the tests, so at the time I was integrating LeakCheck to Apollo tests, I couldn't think of other uses for it. Now that we have them, I propose to make a separate option, might be GRACEFUL_SHUTDOWN for Apollo to consider when it has to terminate a Replica process, and I will add its usage when we have LEACKCHECK=ON. What do you guys think @teoparvanov @yontyon ?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ async def test_wedge_command_and_specific_replica_info(self, bft_network):
         with trio.fail_after(seconds=90):
             done = False
             while done is False:
-                msg = skvbc.get_have_you_stopped_req()
+                msg = skvbc.get_have_you_stopped_req(1)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:131]: Can you use a key-word parameter here? Otherwise `1` looks like some magic number here 😄

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 3]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+        client = bft_network.random_client()
+
+        key, val = await skvbc.write_known_kv()
+
+        await client.write(skvbc.write_req([], [], block_id=0, add_remove_node_command=True))
+
+        with trio.fail_after(seconds=90):
+            done = False
+            while done is False:
+                msg = skvbc.get_have_you_stopped_req(1)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:213]: same here - magic `1`

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 4]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+                rep = await client.read(msg, m_of_n_quorum=bft_client.MofNQuorum.All(client.config, [r for r in range(
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:214]: instead of `[r for r in range(...` can't you just say `bft_network.all_replicas()` ?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 5]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+        await self.validate_stop_on_super_stable_checkpoint(bft_network, skvbc)
+
+        for r in bft_network.all_replicas():
+            last_stable_checkpoint = await self._get_metric(r, bft_network, "Gauges", "lastStableSeqNum")
+            self.assertEqual(last_stable_checkpoint, 300)
+
+        bft_network.stop_all_replicas()
+        # We now expect the replicas to start with a fresh new configuration which means that we
+        # need to see in the logs that isNewStorage() = true. Also,
+        # we expect tp see that lastStableSeqNum = 0 (for example)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:232]: nit - "to see"

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 6]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+        for i in range(100):
+            await skvbc.write_known_kv()
+
+        key, val = await skvbc.write_known_kv()
+
+        live_replicas = bft_network.all_replicas(without=crashed_replicas)
+        # Verify that replicas are now in slow path
+        for r in live_replicas:
+            num_of_slow_path = await self._get_metric(r, bft_network, "Counters", "totalSlowPaths")
+            self.assertGreater(num_of_slow_path, num_of_slow_path_before)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:292]: why do you need to check this explicitly ? It doesn't seem related to your test.

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 7]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+                rsi_rep = client.get_rsi_replies()
+                done = True
+                for r in rsi_rep.values():
+                    if skvbc.parse_rsi_reply(rep, r) == 0:
+                        done = False
+                        break
+
+        for r in live_replicas:
+            last_stable_checkpoint = await self._get_metric(r, bft_network, "Gauges", "lastStableSeqNum")
+            self.assertGreaterEqual(last_stable_checkpoint, 300)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:310]: where does the `300` come from?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 8]
```diff
@@ -220,6 +373,18 @@ async def verify_replicas_are_in_wedged_checkpoint(self, bft_network, previous_c
...
+            while True:
+                with trio.move_on_after(seconds=1):
+                    try:
+                        key = ['replica', mtype, mname]
+                        value = await bft_network.metrics.get(replica_id, *key)
+                    except KeyError:
+                        # metrics not yet available, continue looping
+                        print(f"KeyError! '{mname}' not yet available.")
+                    else:
+                        return value
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_reconfiguration.py:387]: this looks like generic logic that can live in `bft_network.py`

**Review by teoparvanov — COMMENTED** [review]
Some first thoughts... More to come tomorrow probably 😉

File: tests/apollo/CMakeLists.txt (hunk 1 above)
Comment (HristoStaykov) [reply tests/apollo/CMakeLists.txt:75]: Yes, I will prepare the changes, for now `LEAKCHECK=ON` passed as environment variable will only have effect on termination of Replicas.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 9]
```diff
@@ -2777,14 +2796,14 @@ void ReplicaImp::onInfoRequestTimer(Timers::Handle timer) {
 }
 
 void ReplicaImp::onSuperStableCheckpointTimer(concordUtil::Timers::Handle) {
-  if (!isSeqNumToStopAt(lastExecutedSeqNum)) return;
-  if (!checkpointsLog->insideActiveWindow(lastExecutedSeqNum)) return;
-  CheckpointMsg *cpMsg = checkpointsLog->get(lastExecutedSeqNum).selfCheckpointMsg();
+  if (!isSeqNumToStopAt(lastStableSeqNum)) return;
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.cpp:2799]: This wasn't really a bug, because if we stopped on lastExecutedSeqNum then eventually it will become lastStableSeqNum. 
It is just more comprehensive with lastStableCheckpoint

File: tests/apollo/CMakeLists.txt (hunk 1 above)
Comment (yontyon) [reply tests/apollo/CMakeLists.txt]: https://github.com/vmware/concord-bft/pull/884

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 2 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 3 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 4 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 5 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 6 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 7 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: This is the next next checkpoint (the checkpoint we want to stop at).
Anyway, I have added a variable for this

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 8 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 10]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+                   stop_replica_cmd=None,
+                   num_ro_replicas=0)
+        bft_network.change_configuration(conf)
+
+        bft_network.start_all_replicas()
+        for r in bft_network.all_replicas():
+            last_stable_checkpoint = await self._get_metric(r, bft_network, "Gauges", "lastStableSeqNum")
+            self.assertEqual(last_stable_checkpoint, 0)
+
+        await skvbc.assert_kv_write_executed(key, val)
```
[the remark this answers is not in the document]
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 11]
```diff
@@ -47,11 +52,15 @@ class ControlStateManager : public ResPagesClient<ControlStateManager,
   // This struct define the index of each page in this space.
   struct reservedPageIndexer {
     uint32_t update_reserved_page_ = 0;
+    uint32_t set_clean_metadata_reserved_page_ = 1;
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: I have changed the reserved pages mechanism of the control command handler.

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 12]
```diff
@@ -69,5 +78,19 @@ class StopAtNextCheckpointMessage : public concord::serialize::SerializableFacto
   StopAtNextCheckpointMessage(int64_t checkpointToStopAt) : seqNumToStopAt_(checkpointToStopAt){};
   StopAtNextCheckpointMessage() = default;
 };
+
+class SetClearMetadataFlagMessage : public concord::serialize::SerializableFactory<SetClearMetadataFlagMessage> {
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: done

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 13]
```diff
@@ -21,14 +21,19 @@
...
+static constexpr uint32_t ControlHandlerStateManagerNumOfReservedPages = 2;
 class ControlStateManager : public ResPagesClient<ControlStateManager,
                                                   ControlHandlerStateManagerReservedPagesIndex,
                                                   ControlHandlerStateManagerNumOfReservedPages> {
  public:
   void setStopAtNextCheckpoint(int64_t currentSeqNum);
   std::optional<int64_t> getCheckpointToStopAt();
   void clearCheckpointToStopAt();
+
+  void setFlagCleanMetadata(int64_t currentSeqNum);
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: done

File: bftengine/src/bcstatetransfer/BCStateTran.hpp [hunk 14]
```diff
@@ -86,6 +86,8 @@ class BCStateTran : public IStateTransfer {
 
   void addOnTransferringCompleteCallback(std::function<void(uint64_t)>) override;
 
+  void setClearMetadataFlag() override { psd_->setClearDataStoreFlag(); }
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/src/bcstatetransfer/BCStateTran.hpp]: done

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 15]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
         await skvbc.assert_kv_write_executed(key, val)
         await skvbc.write_known_kv()
 
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 7)
+    async def test_remove_nodes(self, bft_network):
```
[the remark this answers is not in the document]
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py:192]: 10x :)

File: tests/apollo/util/bft.py [hunk 16]
```diff
@@ -261,6 +261,34 @@ def existing(cls, config, replicas, clients, client_factory=None):
...
+        such as: n,f,c and the parts that are affected by this change (keys, and clients)
+        We don't want to remove the logs, db and other state related components
+        """
+        # We cannot change anything there are running replicas
+        assert(len(self.procs) == 0)
+
+        # remove all existing clients
+        for client in self.clients.values():
+            client.__exit__()
+        for client in self.reserved_clients.values():
```
[the remark this answers is not in the document]
Comment (yontyon) [reply tests/apollo/util/bft.py:276]: This code was copied from the __exit__ method.
I guess @teoparvanov can answer this.

File: tests/apollo/util/bft.py [hunk 17]
```diff
@@ -261,6 +261,34 @@ def existing(cls, config, replicas, clients, client_factory=None):
         bft_network._init_metrics()
         return bft_network
 
+    def change_configuration(self, config):
+        """
+        When changing an existing bft-network, we would want to change only its configuration related parts
+        such as: n,f,c and the parts that are affected by this change (keys, and clients)
+        We don't want to remove the logs, db and other state related components
+        """
+        # We cannot change anything there are running replicas
```
[the remark this answers is not in the document]
Comment (yontyon) [reply tests/apollo/util/bft.py]: done

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 18]
```diff
@@ -187,6 +187,159 @@ async def test_semi_manual_upgrade(self, bft_network):
...
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 7)
+    async def test_remove_nodes_with_f_failures(self, bft_network):
+        """
+        In this test we show how a system operator can remove nodes (and thus reduce the cluster) from 7 nodes cluster
+        to 4 nodes cluster even when f nodes are not responding
+        For that the operator performs the following steps:
+        1. Stop 2 nodes (f=2)
+        2. Send a remove_node command - this command also wedges the system
+        3. Verify that all live (including the removed candidates) have stopped
```
[the remark this answers is not in the document]
Comment (yontyon) [reply tests/apollo/test_skvbc_reconfiguration.py]: done

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 19]
```diff
@@ -20,19 +20,38 @@
 
 namespace bftEngine {
 
+class ControlStatePage : public concord::serialize::SerializableFactory<ControlStatePage> {
+ public:
+  int64_t seqNumToStopAt_ = 0;
+  int64_t eraseMetadataAtSeqNum_ = 0;
+  ControlStatePage() {
+    static_assert(sizeof(ControlStatePage) < 4096, "The page exceeds the maximal size of reserved page");
```
Comment (andrewjstone) [inline bftengine/include/bftengine/ControlStateManager.hpp:28]: Nice :)

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 20]
```diff
@@ -29,37 +29,60 @@ void ControlStateManager::setStopAtNextCheckpoint(int64_t currentSeqNum) {
...
-  try {
-    concord::serialize::Serializable::deserialize(inStream, msg);
-  } catch (std::exception& e) {
-    LOG_WARN(GL, e.what());
-    return {};
-  }
-  if (msg.seqNumToStopAt_ < 0) {
+  concord::serialize::Serializable::deserialize(inStream, page_);
+  if (page_.seqNumToStopAt_ == 0) return {};
+  if (page_.seqNumToStopAt_ < 0) {
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ControlStateManager.cpp:48]: This seems like a fatal error. I left a similar comment on the previous review. Do you agree?

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 21]
```diff
@@ -29,37 +29,60 @@ void ControlStateManager::setStopAtNextCheckpoint(int64_t currentSeqNum) {
...
+  if (!enabled_) return {};
+  if (page_.eraseMetadataAtSeqNum_ != 0) return page_.eraseMetadataAtSeqNum_;
+  if (!state_transfer_->loadReservedPage(resPageOffset(), sizeOfReservedPage_, scratchPage_.data())) {
+    return {};
+  }
+  std::istringstream inStream;
+  inStream.str(scratchPage_);
+  concord::serialize::Serializable::deserialize(inStream, page_);
+  if (page_.eraseMetadataAtSeqNum_ == 0) return {};
+  if (page_.eraseMetadataAtSeqNum_ < 0) {
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ControlStateManager.cpp:81]: Same as above. This is a fatal error. You never write a sequence number less than 0, so this implies corruption.

**Review by andrewjstone — COMMENTED** [review]
@yontyon This looks much better, especially the ControlStatePage serialiization! Thanks for all the good work. 

I have couple comments I think you still need to look into though:
 * Fatal errors for deserialization invariant violations
 * Reconfig without N-N quorum

Please take a look. I'm happy to discuss with you offline as well.

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 20 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp:48]: Agreed

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 21 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp:81]: Agreed.
10x!

**Comment by yontyon** [comment]
> @yontyon This looks much better, especially the ControlStatePage serialiization! Thanks for all the good work.
> 
> I have couple comments I think you still need to look into though:
> 
> * Fatal errors for deserialization invariant violations
> * Reconfig without N-N quorum
> 
> Please take a look. I'm happy to discuss with you offline as well.

As for reconfiguring without N/N, this is essential for removing nodes because the major reason for removing a node is if this node permanently crashed.
Now, as we assume that we will use this feature only when we know in advance that some nodes are down, it is safe enough.
For now, we won't use this feature for other reconfiguration actions (such as upgrade) until we will have a more profound discussion also with @ittaiab

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 21 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp:80]: Agreed
10x!

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 22]
```diff
@@ -20,6 +20,27 @@
 
 namespace bftEngine {
 
+class ControlStatePage : public concord::serialize::SerializableFactory<ControlStatePage> {
+ public:
+  int64_t seqNumToStopAt_ = 0;
```
Comment (robem) [inline bftengine/include/bftengine/ControlStateManager.hpp:25]: Please change member variables to snake_case. Similar, function parameters. https://google.github.io/styleguide/cppguide.html#Variable_Names

**Review by robem — CHANGES_REQUESTED** [review]

**Comment by andrewjstone** [comment]
> > @yontyon This looks much better, especially the ControlStatePage serialiization! Thanks for all the good work.
> > I have couple comments I think you still need to look into though:
> > 
> > * Fatal errors for deserialization invariant violations
> > * Reconfig without N-N quorum
> > 
> > Please take a look. I'm happy to discuss with you offline as well.
> 
> @andrewjstone As for reconfiguring without N/N, this is essential for removing nodes because the major reason for removing a node is if this node permanently crashed.
> Now, as we assume that we will use this feature only when we know in advance that some nodes are down, it is safe enough.
> For now, we won't use this feature for other reconfiguration actions (such as upgrade) until we will have a more profound discussion also with @ittaiab

I agree with this assessment. Can you please change some of the wording in comments though to reflect that N-F only currently works for replacing and removing nodes.

File: bftengine/include/bftengine/Replica.hpp [hunk 23]
```diff
@@ -78,6 +79,12 @@ class IReplica {
   using IReplicaPtr = std::unique_ptr<IReplica>;
   static IReplicaPtr createNewReplica(
       ReplicaConfig *, IRequestsHandler *, IStateTransfer *, bft::communication::ICommunication *, MetadataStorage *);
+  static IReplicaPtr createNewReplica(ReplicaConfig *,
+                                      IRequestsHandler *,
+                                      IStateTransfer *,
+                                      bft::communication::ICommunication *,
+                                      MetadataStorage *,
+                                      bool &);
```
Comment (andrewjstone) [inline bftengine/include/bftengine/Replica.hpp:87]: Can you please give a name to this bool? Better yet make it a concrete type (enum). Overloading a bool parameter makes calling code difficult to understand.

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 24]
```diff
@@ -190,6 +196,14 @@ IReplica::IReplicaPtr IReplica::createNewReplica(ReplicaConfig *replicaConfig,
   return replicaInternal;
 }
 
+IReplica::IReplicaPtr IReplica::createNewReplica(ReplicaConfig *replicaConfig,
+                                                 IRequestsHandler *requestsHandler,
+                                                 IStateTransfer *stateTransfer,
+                                                 bft::communication::ICommunication *communication,
+                                                 MetadataStorage *metadataStorage) {
+  bool dummy;
```
Comment (andrewjstone) [inline bftengine/src/bftengine/BFTEngine.cpp:204]: What does this signify? As in prior comment it'd be better to pass an enum to the constructor instead of a bool.

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 20 above)
Comment (andrewjstone) [reply bftengine/src/bftengine/ControlStateManager.cpp:48]: You also need to actually call std::terminate after this log message.

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 21 above)
Comment (andrewjstone) [reply bftengine/src/bftengine/ControlStateManager.cpp:80]: Same as above. You need to call std::terminate after any fatal error.

File: bftengine/src/bftengine/PersistentStorageImp.cpp [hunk 25]
```diff
@@ -905,6 +914,20 @@ bool PersistentStorageImp::nonExecSetIsAllowed() {
   return setIsAllowed() &&
          (!hasDescriptorOfLastExecution() || descriptorOfLastExecution_.executedSeqNum <= lastExecutedSeqNum_);
 }
+void PersistentStorageImp::setEraseMetadataStorageFlag() {
+  bool eraseMtOnStartUp = true;
```
Comment (andrewjstone) [inline bftengine/src/bftengine/PersistentStorageImp.cpp:918]: spell out whatever Mt is ? Presumably metadata :)

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 26]
```diff
@@ -2500,6 +2500,25 @@ void ReplicaImp::onSeqNumIsStable(SeqNum newStableSeqNum, bool hasStateInformati
...
     tryToSendPrePrepareMsg();
   }
+
+  auto seq_num_to_stop_at = controlStateManager_->getCheckpointToStopAt();
+  if (seq_num_to_stop_at.has_value() && seq_num_to_stop_at.value() == newStableSeqNum) {
+    LOG_INFO(GL,
+             "Informing control state manager that consensus should be stopped (without n/n replicas): " << KVLOG(
+                 newStableSeqNum, metric_last_stable_seq_num_.Get().Get()));
+    if (getRequestsHandler()->getControlHandlers()) {
+      getRequestsHandler()->getControlHandlers()->onStableCheckpoint();
```
[the remark this answers is not in the document]
Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaImp.cpp:2515]: As mentioned in another comment, I would add a comment above this whole clause, to say that this is only safe for removing or replacing nodes.

**Review by andrewjstone — CHANGES_REQUESTED** [review]

File: bftengine/src/bftengine/BFTEngine.cpp (hunk 24 above)
Comment (yontyon) [reply bftengine/src/bftengine/BFTEngine.cpp:204]: I didn't want to break the current API and implementation, so I have added another signature with the reference to bool in which we return if the metadata storage has been deleted on startup.

File: bftengine/include/bftengine/Replica.hpp (hunk 23 above)
Comment (yontyon) [reply bftengine/include/bftengine/Replica.hpp:87]: I would prefer to keep using bool due to code simplicity

File: bftengine/src/bftengine/PersistentStorageImp.cpp (hunk 25 above)
Comment (yontyon) [reply bftengine/src/bftengine/PersistentStorageImp.cpp:918]: fixed

File: bftengine/include/bftengine/ControlStateManager.hpp (hunk 22 above)
Comment (yontyon) [reply bftengine/include/bftengine/ControlStateManager.hpp]: fixed

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 21 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: fixed

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 20 above)
Comment (yontyon) [reply bftengine/src/bftengine/ControlStateManager.cpp]: fixed

File: bftengine/include/bftengine/Replica.hpp (hunk 23 above)
Comment (andrewjstone) [reply bftengine/include/bftengine/Replica.hpp]: No problem.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 27]
```diff
@@ -2505,6 +2505,33 @@ void ReplicaImp::onSeqNumIsStable(SeqNum newStableSeqNum, bool hasStateInformati
   if (!oldSeqNum && currentViewIsActive() && (currentPrimary() == config_.replicaId) && !isCollectingState()) {
     tryToSendPrePrepareMsg();
   }
+
+  auto seq_num_to_stop_at = controlStateManager_->getCheckpointToStopAt();
+
+  // Below we handle the case of removing a node. Note that when removing nodes we cannot assume we will have n/n
+  // checkpoint because some of the replicas may not be responsive. For that we also mark that we got to a stable (n-f)
+  // checkpoint.
+  // Note: Currently this is not sage to rely on this n-f wedged checkpoint on any other reconfiguration action except
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.cpp:2514]: typo: sage -> safe

Comment (andrewjstone) [reply bftengine/src/bftengine/ReplicaImp.cpp:2514]: except of -> except for

**Review by andrewjstone — APPROVED** [review]
Looks great @yontyon :) Thanks for taking the time to address all the feedback from multiple reviews. It's awesome to have this functionality!
